src/main/java/org/openhab/binding/zigbee/converter/ZigBeeConverterColorColor.java

```diff
@@ -106,72 +266,186 @@ public void run() {
+                    updateBrightness(brightness);
+                } else if (attribute.getCluster().getId() == ZclColorControlCluster.CLUSTER_ID
+                        && attribute.getId() == ZclColorControlCluster.ATTR_CURRENTHUE) {
+                    Integer value = (Integer) attribute.getLastValue();
+                    float hue = value * 360.0f / 254.0f;
+                    if (hue != lastHue) {
+                        lastHue = hue;
+                        hueChanged = true;
+                    }
+                } else if (attribute.getCluster().getId() == ZclColorControlCluster.CLUSTER_ID
+                        && attribute.getId() == ZclColorControlCluster.ATTR_CURRENTSATURATION) {
+                    Integer value = (Integer) attribute.getLastValue();
+                    float saturation = value * 100.0f / 254.0f;
+                    if (saturation != lastSaturation) {
+                        lastSaturation = saturation;
+                        saturationChanged = true;
+                    }
+                } else if (attribute.getCluster().getId() == ZclColorControlCluster.CLUSTER_ID
+                        && attribute.getId() == ZclColorControlCluster.ATTR_CURRENTX) {
+                    Integer value = (Integer) attribute.getLastValue();
+                    float x = value / 65536.0f;
+                    if (x != lastX) {
+                        lastX = x;
+                        xChanged = true;
+                    }
+                } else if (attribute.getCluster().getId() == ZclColorControlCluster.CLUSTER_ID
+                        && attribute.getId() == ZclColorControlCluster.ATTR_CURRENTY) {
+                    Integer value = (Integer) attribute.getLastValue();
+                    float y = value / 65536.0f;
+                    if (y != lastY) {
+                        lastY = y;
+                        yChanged = true;
+                    }
                 }
-            }
-            if (attribute.getId() == ZclColorControlCluster.ATTR_CURRENTSATURATION) {
-                Integer value = (Integer) attribute.getLastValue();
-                if (value != null) {
-                    PercentType saturation = new PercentType(value * 100 / 254);
-                    currentHSB = new HSBType(currentHSB.getHue(), saturation, currentHSB.getBrightness());
-                    updateChannelState(currentHSB);
+                if (hueChanged && saturationChanged) {
+                    updateColorHSB();
+                } else if (xChanged && yChanged) {
+                    updateColorXY();
+                } else if (hueChanged || saturationChanged || xChanged || yChanged) {
+                    // Wait some time and update anyway if only one attribute in each pair is updated
+                    new Thread() {
+                        @Override
+                        public void run() {
+                            try {
+                                sleep(500);
```

Comment:
Not really.

Within the relevant synchronized portions of code, the [hue, saturation, x, y]Changed set of variables are set to false when the update happens, so after the second attribute is received, the delayed thread will see them as false and just skip the update. Normally it will be the second attribute in each pair forcing the update inmediately, unless the attribute is reported 500ms later (this is just a safeguard in case only one is updated). 500ms seems a good compromise, as both attributes should be reported within this timeframe (in my IKEA lights even both are reported in the same message).

Reply:
I suspect this isn't a very good way to acheive what you want. You will potentially create a lot of threads if the attributes update often (which might happen when someone is playing around with settings). I would suggest to use a ScheduledExecutorService - create this in your initialisation method, then when you want to create a new timer here, you can do so and can cancel anything that is already waiting to run.

```
        ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor();

        ScheduledFuture<?> timer = scheduler.schedule(task, 500, TimeUnit.MILLISECONDS);
        timer.cancel(true);
```

What do you think?

Reply:
Done & testing